arborist/resource.go

```diff
@@ -30,6 +31,43 @@ type ResourceOut struct {
 	Subresources []string `json:"subresources"`
 }
 
+var regPercent *regexp.Regexp = regexp.MustCompile(`%`)
+var regSlashEncoded *regexp.Regexp = regexp.MustCompile(`%2F`)
+
+func UnderscoreEncode(decoded string) string {
+	// Per https://www.ietf.org/rfc/rfc3986.txt, there are the following
+	// unreserved characters (aside from alphanumeric and underscore) which do
+	// not urlencode:
+	//   - `-`
+	//   - `.`
+	//   - `~`
+	// Percent encoding uses hexadecimal, which we mustn't conflict with. So we
+	// use some made-up codes for these which will not overlap.
```

Comment:
what happens if you encode _ - . ~ using the percent hexadecimal encoding?

Reply:
That's what [these](https://github.com/uc-cdis/arborist/blob/558f563d7de8d8233f15f6c991c1ea5ec18a9973/arborist/resource.go#L47-L50) are for.
```
fmt.Println(UnderscoreEncode("_-.~"))

_S0_S1_S2_S3
```

Reply:
i mean, why `hexadecimal, which we mustn't conflict with`? why not use `_7E` for `~` instead of `_S3`?

Reply:
I suppose you could—I might not have a complete understanding—according to [RFC 3986](https://tools.ietf.org/html/rfc3986#section-2.3) those characters specifically are "unreserved", meaning they wouldn't need to be encoded ordinarily (they aren't affected by `url.QueryEscape` for instance). So I wasn't sure about the treatment for those in encoding/decoding in general. You could use `_7E` instead, I think, I just picked something completely different to guarantee it wouldn't collide.